src/lib/syncable_ledger/syncable_ledger.ml

```diff
@@ -221,55 +282,53 @@ end = struct
 
   type query = Addr.t Query.t
 
+  (* Provides addresses at an specific depth from this address *)
+  let rec intermediate_range : index -> Addr.t -> index -> Addr.t list =
+   fun ledger_depth addr i ->
+    match i with
+    | 0 ->
+        [ addr ]
+    | i ->
+        let left, right =
+          (* TODO: may be better to propagate the error *)
+          Option.value_exn
+            ( Or_error.ok
+            @@ Or_error.both
+                 (Addr.child ~ledger_depth addr Direction.Left)
+                 (Addr.child ~ledger_depth addr Direction.Right) )
+        in
+        let left = intermediate_range ledger_depth left (i - 1) in
+        let right = intermediate_range ledger_depth right (i - 1) in
+        left @ right
```

Comment:
Doing this list concatenation results in `n^2` allocations. We also immediately convert it to an array immediately later. I'd recommend using an `Array.init` and appending the index, something like [this](https://github.com/MinaProtocol/mina/commit/3e7b340289).